core/src/main/java/com/datastax/oss/driver/internal/core/util/concurrent/Reconnection.java

```diff
@@ -229,7 +244,7 @@ private void onNextAttemptCompleted(Boolean success, Throwable error) {
         reallyStop();
       } else {
         assert state == State.ATTEMPT_IN_PROGRESS;
-        scheduleNextAttempt();
+        scheduleNextAttempt(error);
```

Comment:
This assumes that `reconnectionTask` will rethrow errors, which is not the case currently because the previous convention was that tasks would handle errors on their side an only return a success/failure boolean. So only unexpected errors will be passed here; see for example how `ChannelPool.onAllConnected()` swallows authentication errors, we want them to be reported to the policy now.

I think we need to change `reconnectionTask` to `Callable<CompletionStage<Void>>`: a successful stage means stop reconnecting, and failed stage means pass the error to the policy (which is already done here, we just need to remove the warning about unexpected errors).
Then `ChannelPool.onAllConnected()` can simply rethrow errors, and `ControlConnection.reconnect` will do:
```java
connect (
  ...
  error -> result.completeExceptionally(error));
```

Reply:
I got most of the way through this... honestly I think this is the wrong way to solve this problem. Not the reconnectionTask in general, but the entire idea of delegating when to retry to the underlying policy.

Much of the validation of when to retry and when not to is getting pushed entirely down into the retry policies next delay()... this leaves quite a bit discretion in the retry policy, and also complicates the implementation of the retry policy as they need to know the intricacies of each scenario. I don't think it's realistic for a end user to be able to extend retry policy in this manner outside of our team, and I think I'm likely to introduce a bunch of bugs in the process. It also invites duplication of code.. since the retry policies scheduler is a private inner class, which make inheritance a little tricky.

I talked to a member of the evangelist team. I think they would be satisfied simply by having the log level increased in the cases of authentication failure. This would be far easier way to solve their problem.

Reply:
This does seem to be a bit cumbersome to have to propagate the connection Exceptions down to the ReconnectionPolicy in order to determine if we should abort the reconnection attempt(s). And I agree that end users trying to provide a custom implementation for the policy will have a difficult time if they aren't aware of all the reasons a connection attempt would fail and how they would manifest themselves in the code/Exceptions.
However, if all we do is increase the log level, the only thing that does is make it a little more obvious (if you are looking at the logs) that the client can't connect because it can't authenticate, correct? It will still keep retrying, and the log will be the only indication that it's not hung.